up to date document workloads/templates/vm-creation.md (#150)
kubevirt/user-guide: workloads/templates/vm-creation.md

These templates provide a combination of some key factors that could be further customized and processed to have a Virtual Machine object.

The key factors which define a template are
 - Workload
  Most Virtual Machine should be *generic* to have maximum flexibility; the *highperformance* workload trades some of this flexibility to
  provide better performances.
- Guest Operating System (OS)
  This allow to ensure that the emulated hardware is compatible with the guest OS. Furthermore, it allows to maximize the stability
  of the VM, and allows performance optimizations.
- Size (flavor) 
  Defines the amount of resources (CPU, memory) to allocate to the VM.

## WebUI
Kubevirt project has [the official UI](https://github.com/kubevirt/web-ui). This UI supports creation VM using templates and templates features - flavors and workload profiles.
To create VM from template, choose WorkLoads in the left panel >> press to the "Create Virtual Machine" blue button >> choose "Create from Wizzard". Next, you have to see "Create Virtual Machine" window

> ## common-templates

Reviewer: ```suggestion
## Common templates
```

> There is the [common-templates subproject](https://github.com/kubevirt/common-templates/) subproject. It provides official prepaired and useful templates. [Additional doc available](/workloads/templates/common-templates.md). You can also create templates by hand. You can find an example below, in the "Example template" section.

Reviewer: ```suggestion
Kubevirt provides a set of tested useful templates via the [common-templates](https://github.com/kubevirt/common-templates/) subproject. Those templates also contain additional metadata that provide more information to the WebUI that can then offer improved selection and validation logic. [Additional doc available](/workloads/templates/common-templates.md).

We recommend using a provided template, however you can create custom templates by hand too if needed. Please find an example below, in the "Example template" section.
```

Author: I've added it in the latest commit @MarSik

## Overview

The KubeVirt projects provides a set of [templates](https://docs.okd.io/latest/dev_guide/templates.html) to create VMs to handle common usage scenarios.
These templates provide a combination of some key factors that could be further customized and processed to have a Virtual Machine object.

The key factors which define a template are
 - Workload
  Most Virtual Machine should be *generic* to have maximum flexibility; the *highperformance* workload trades some of this flexibility to
  provide better performances.
- Guest Operating System (OS)
  This allow to ensure that the emulated hardware is compatible with the guest OS. Furthermore, it allows to maximize the stability
  of the VM, and allows performance optimizations.
- Size (flavor) 
  Defines the amount of resources (CPU, memory) to allocate to the VM.

## WebUI
Kubevirt project has [the official UI](https://github.com/kubevirt/web-ui). This UI supports creation VM using templates and templates features - flavors and workload profiles.

> To create VM from template, choose WorkLoads in the left panel >> press to the "Create Virtual Machine" blue button >> choose "Create from Wizzard". Next, you have to see "Create Virtual Machine" window

Reviewer: ```suggestion
To create VM from template, choose _WorkLoads_ in the left panel, press the _Create Virtual Machine_ blue button, choose _Create from Wizard_. Next, you will see the "Create Virtual Machine" window and you should be able to select the guest operating system, workload profile and size.
```

The spelling and formatting was really bad, you should not change styles within one sentence.

Author: I've add it in the latest commit
>The spelling and formatting
Sorry

> In order to create a virtual machine via OpenShift CLI, you need to provide a template defining the corresponding object and its metadata.

Reviewer: This whole thing needs an update. We already ship the templates the user can use. Francesco's patch describes the format https://github.com/kubevirt/user-guide/pull/151 and the templates are released here https://github.com/kubevirt/common-templates/releases

So the user does not have to create their own template as described below, they can take one of the templates we released for them. I know you mention it at the bottom, but we should place it more prominently so people actually notice. Otherwise nobody will use them.

Author: Maybe at least good idea to mentioned regarding cli opportunity? I was inspired by OpenShift/Origin guides.

Reviewer 2: Martin raised a good point. I think you can maybe just add references to the openshift docs, no need to replicate it there.

apiVersion: v1
kind: Template
metadata:
  annotations:
    description: OCP KubeVirt Fedora 27 VM template
    iconClass: icon-fedora
    tags: kubevirt,ocp,template,linux,virtualmachine

>   creationTimestamp: null

Reviewer: ```suggestion
```

Reviewer: There is really no need to add null elements to the template. Or is there?

Author: In this case also good idea to send MR to change files:

```bash
workloads/controllers/virtual-machine.md:123:  creationTimestamp: null
workloads/controllers/virtual-machine.md:131:      creationTimestamp: null
workloads/virtual-machines/unresponsive-nodes.md:22:    creationTimestamp: 2018-11-05T08:55:53Z
workloads/virtual-machines/image-upload.md:88:  creationTimestamp: null
workloads/virtual-machines/disks-and-volumes.md:399:  creationTimestamp: null
workloads/virtual-machines/disks-and-volumes.md:514:  creationTimestamp: null
workloads/virtual-machines/disks-and-volumes.md:594:  creationTimestamp: null
workloads/virtual-machines/disks-and-volumes.md:655:  creationTimestamp: null
workloads/virtual-machines/disks-and-volumes.md:780:  creationTimestamp: null
workloads/virtual-machines/disks-and-volumes.md:868:  creationTimestamp: null
workloads/virtual-machines/disks-and-volumes.md:1030:  creationTimestamp: null
```

Author: @MarSik ?

Reviewer: Sure why not, but it is not a part of this PR, just do not introduce any more..

  labels:
    kubevirt.io/os: fedora27
    miq.github.io/kubevirt-is-vm-template: "true"
  name: vm-template-fedora
objects:
- apiVersion: kubevirt.io/v1alpha2
  kind: VirtualMachine
  metadata:

>     creationTimestamp: null

Reviewer: ```suggestion
```

Author: Please check comment above

    labels:
      kubevirt-vm: vm-${NAME}
      kubevirt.io/os: fedora27
    name: ${NAME}
  spec:
    running: false
    template:
      metadata:

>         creationTimestamp: null

Reviewer: ```suggestion
```

Author: Please check comment above

```yaml
---
kind: PersistentVolume
apiVersion: v1
metadata:
  name: mypv
  labels:
    type: local
spec:
  storageClassName: manual
  capacity:
    storage: 10G
  accessModes:
    - ReadWriteOnce
  hostPath:
    path: "/mnt/sda1/images/testvm"
---
kind: PersistentVolumeClaim
apiVersion: v1
metadata:
  name: mypvc
spec:
  storageClassName: manual
  accessModes:
    - ReadWriteOnce
  resources:
    requests:
      storage: 10G

```

If you create this PV/PVC, then you have to put VM image in the file path

```bash
/mnt/sda1/images/testvm/disk.img
```

Avaible in the each OpenShift/Kubevirt compute nodes.

## WebUI
Kubevirt project has [the official UI](https://github.com/kubevirt/web-ui). This UI supports creation VM using templates and templates features - flavors and workload profiles.
To create VM from template, choose WorkLoads in the left panel >> press to the "Create Virtual Machine" blue button >> choose "Create from Wizzard". Next, you have to see "Create Virtual Machine" window

> ## common-templates

Reviewer: As I mentioned above, this should be much more visible I think. The user should prefer using the pre-created templates instead of writing his own.

Author: >Maybe move this up in the document?
Agree

Author: Moved

Reviewer: Not really, you moved the UI section, not the template section. What I would do is moving the common-templates section to the top and then below it writing something like: "You can also create templates by hand and here is an example of how it could look like" and then I would use the existing Example template section

``` bash
$ oc patch virtualmachine testvm --type merge -p '{"spec":{"running":true}}'
virtualmachine.kubevirt.io/testvm patched
```

Do not forget about virtctl tool. Use it in the real cases instead of using kubernetes API, can be more convinient. Example: 

```bash
$ virtctl start testvm
VM testvm was scheduled to start
```

> As fast as VM starts, kubernates creates new type of object - VirtualMachineInstance. It has similar name as VirtualMachine. Example (not full output, it's too big):

Reviewer 2: minor: "output trimmed" ?

## Using registry images

Kubevirt VM templates, just like kubevirt VM/VMI yaml configs, supports creating VM's disks from registry. RegistryDisk is a special type volume which supports downloading images from user-defined registry server.

## **Hack** - use pre-downloaded image

> Kubevirt VM templates, just like kubevirt VM/VMI yaml configs, can use pre-downloaded VM image, which can be a useful feature especially in the debug/development/testing cases. No special parameters required in the VM template or VM/VMI yaml config. The main idea is to create Kubernetes PersistentVolume and PersistentVolumeClaim corresponding to existing image in the file system. Example:

Reviewer 2: is required?
